pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java

```diff
@@ -640,8 +643,22 @@ public synchronized void setupBrokerPublishRateLimiterMonitor() {
         }
     }
 
-    @Override
     public void close() throws IOException {
+        try {
+            closeAsync().get();
+        } catch (ExecutionException e) {
+            if (e.getCause() instanceof IOException) {
+                throw (IOException) e.getCause();
+            }
+            throw new PulsarServerException(e.getCause());
+        } catch (IOException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new PulsarServerException(e);
+        }
+    }
+
+    public CompletableFuture<Void> closeAsync() throws IOException {
```

Comment:
here we have the same problem, a method that returns CompletableFuture should not "throw" exceptions.
Because the caller is expected to handle errors using the CompletableFuture API

Reply:
resolved in recent changes